mini-projects/00-jq-rs/src/filters.rs

```diff
@@ -0,0 +1,499 @@
+use core::panic;
+use serde_json::Value;
+
+use clap::{Arg, Command};
+use std::env;
+use std::fs::File;
+use std::io::BufReader;
+
+#[derive(Debug, PartialEq)]
+pub enum Filter {
+    Identity,
+    ObjectIdentifier(String),
+    ArrayIndex(isize),
+    ArraySlice(Option<isize>, Option<isize>),
+    ArrayIterator,
+    Add,
+    Length,
+    Del(Box<Filter>),
+    // Other filters as needed
+}
+
+//name:
+//obj_identify -- .<KEY_NAME>
+//array_index --  .[i32]
+//array_slice -- .[<start_number>:<end_number>]
+//pipe -- pass output of one filter to another
+// -- handle in main loop?
+//array_iterator -- turn an array into an iterator
+// -- effectively expands an array into multiple objects
+
+//add -- adds all values in an array
+//length -- returns the length of the input
+//del -- del(.[<indexes>]) -- return everything but deleted
+
+pub fn apply_filters(values: Vec<Value>, filters: &[Filter]) -> Result<Vec<Value>, String> {
+    let mut current_values = values;
+
+    for filter in filters {
+        let mut next_values = Vec::new();
+        for value in current_values {
+            let result = apply_single_filter(value, filter)?;
+            next_values.extend(result);
+        }
+        current_values = next_values;
+    }
+
+    Ok(current_values)
+}
+
+pub fn apply_single_filter(value: Value, filter: &Filter) -> Result<Vec<Value>, String> {
+    match filter {
+        Filter::Identity => Ok(vec![value]),
```

Comment:
similar to my previous comment about how you handle the array iterators, this line to me seems silly to need to wrap the value in a vec, just for you to call extend it in the higher function